### ISSUE TITLE
Review Maintenance Framework

### ISSUE THREAD

**Issue by Chingujo** [department-of-veterans-affairs/caseflow-efolder#963]
Created a maintenance frame work that can be found [here](https://github.com/department-of-veterans-affairs/appeals-design-research/blob/master/Projects/eFolder%20Express/Maintenance%20Framework%20(eFolder%20Express).MD). 

@lowellrex can you please review and provide feedback/concerns on the proposed model for how we maintain along with requirements. Thanks. 

# Maintenance Framework for eFolder Express

**Context**:
With the test release of the React release largely successful pilot 1 and pending pilot 2, we will be releasing the full react application to all eFolder Express users within the next few weeks. This calls into question how to maintain this product in the immediate and long term. This document outlines structure, process, responsibilities, and requirements. 

#### Team 
Product Owner: Phillip Jo
Dev lead: Lowell Wood

### Structure
Learning from Caseflow certification - another product in maintenance for the past three to four months - the structure has been that the primary dev and product owner transitions into maintaining the health of a product. To ensure that we are proactively thinking about product health versus reacting to production issues, the product owner and dev lead will have an **_optional and as needed only_** meeting as a placeholder to discuss the following as necessary: 

- eFolder express performance, health, and status of any major bug changes
- design and ticket changes 
- access requests and affects on system load 
- any notable product changes

### Responsibilities 
- Product owners will continue to conduct the following: 
  - Stakeholder management including access requests, new product and initiative developments, and internal product changes 
  - Working with users to field and triages bugs and correctly prioritizng them to work
  - Maintain the design and UX of the product and priortizing them accordingly
  - Working with support to communicate with users regarding versioning changes and releases after major outages
  - Working with support to priortize and answer feature requests

- Dev lead
  - Maintains the technical health of the product monitoring important library updates, major bug fixes that affect application functionality, and
    any major changes in regards to other products that use Ex such as Reader
  - Works with product owner to priortize and implement design changes
  - Monitoring alerts for the app to priortize fixes with PO
 
 
  
### To be completed before maintenance
- [ ] Successful roll out of react application for all users and touch ups of production bugs 
- [ ] Fixing of issues related to [large files](https://github.com/department-of-veterans-affairs/caseflow-efolder/issues/854). We need a sustainable fix for this: 
     - Right now, for big files, we need a MAC user to download the large zip file and split the file into 2 gb chunks to send to the user. 
     - It takes about an hour to do this for support staff and is not sustaniable
     - The frequency seems to be about 4-5 a month

#### High priorty (O crap fix asap)
- Major bugs: Any bugs or changes that cause an outage or degraded performance for users with eFolder Express and across products
- Major design bugs: Any major design bugs that drastically affect the user experience. For example, the document table no longer rendering or rendering incorrectly. Drastic to be determined by Product owner and Dev lead. 

#### Medium to low priority (to the backlog)
- Any tech and design updates that are nice to haves and exist in the tech backlog as bandwidth allows: 
  - Minor design changes 
  - Non essential library updates 
  - Feature requests
  - Copy changes 
- To be triaged accordingly with product owner and dev lead based on bandwidth

**Comment by lowellrex** [comment]
Looks great! Two thoughts:

## Ongoing dependency upgrades

What is our plan for upgrading dependencies (specifically ruby gems)? I see two distinct options:
1. Regularly (monthly?) check our gems for updates, install those updates (only point releases? all releases?), and confirm that our application still behaves as expected.
1. Only upgrade gems when they are affected by known vulnerabilities.

The first option will requires some amount of ongoing developer time (4 hours a month?), but ensures that we never fall too far behind the current version of any of our dependencies and allows us to proactively document when we decide not to upgrade a third-party dependency (for instance, if doing so will force us to change a significant amount of the application).

The second option incurs no ongoing developer effort, however since I assume we will want to patch any known vulnerabilities, we will still need to occasionally upgrade our dependencies (this has happened not [once](https://github.com/department-of-veterans-affairs/caseflow-efolder/pull/846), but [twice](https://github.com/department-of-veterans-affairs/caseflow-efolder/pull/950) in the past 3 months). Consider the case where we need to upgrade a dependency to respond to a known vulnerability but haven't upgraded the dependency in question for 9 months. We will be forced to make sure that all of the updates in the intervening period don't break the application (and fix the ways those upgrades do break our application if they do) before we can roll out the vulnerability fix.

Either way, we need to make sure we have some automated alerting in place for vulnerabilities in our dependencies. Currently, the primary way we discover dependency vulnerabilities has been through `bundle-audit` which only runs when we run our continuous integration test suite. Since we will be doing that far less frequently we need to make sure we have some other alerting mechanism in place. I think github has a tool for this (that we are perhaps already using?), I will investigate further.

## Ongoing time commitment for support staff

I think it might be worthwhile to establish some level of acceptable time commitment for support staff to dedicate to efolder express issues before requiring a fix to be developed and deployed. Is it alright for support staff to dedicate 5 hours a month to a process that solves a bug in the application? 20 hours?

I think this might be worth determining upfront because the development cycle for even seemingly trivial bug fixes could be quite dramatic as developers get further away from having an active working context of the application (and introducing new code also opens up the possibility of introducing new bugs).

**Comment by Chingujo** [comment]
@lowellrex 

Ace thoughts and feedback as always. Some thoughts and questions below =D

## Re: Ongoing dependency upgrades
Great analysis of the tech trad offs here and will defer this to you on the final say. A few questions: 

- Could we take a middle path or will the amount of time spend be the same or less? For example, if we do this monthly and spend 4 hours a month will it be 12 hours if we do this every 3 months or will it be more than that due to catching up on dependencies and context?

- Does upgrading dependencies cause any conflict across applications? For example, does it matter if we are on a different version than another products that share dependencies with eFolder Express? 

## Re: Ongoing time commitment for support staff
Totally agreed for small issues and non important issues. 

In regards to the corrupted zip file with big files the time saved versus allotted is the big question here. For example, we had three to four of these issues in the past month and if we assume that cadence takes about 2 hours each at 6-8 hours a month X 12, it can add up pretty quickly. So: 

- Wondering if it is worth fixing this or at least doing an initial investigation. If this fix takes an unrealistic amount of dev time amiss other priorities, we should set that expectation with support. The concern is that we may see an increase volume in the amount of issues as we roll out to more users. My preference is toward a more sustainable solution

- My assumption here is in line with yours that it is easier to fix now when we have developers with context than when we don't.

**Comment by lowellrex** [comment]
*I don't think the following is much of a concern at all, I just want to document it here for reference in case we end up having a conversation along these lines later.*

One other thought on the technical side is that we might have long running efolder express EC2 instances. Currently, we create new EC2 instances every time we deploy new efolder code and every time we [scale up from 2 to 7 instances](https://github.com/department-of-veterans-affairs/appeals-deployment/blob/bddd7ec2c54184d26262e2f572fa0657d04c9404/ansible/utility-roles/aws-deploy/files/schedule_scale_down.py#L84) every morning. Since we will no longer be deploying code regularly, the only time we will be deploying new instances is when we scale up each morning.

The auto scaling group logs indicate that when we scale up from 2 to 7 instances each morning we only add the 5 new instances, and don't replace the 2 existing instances. This opens up the possibility that we could have as many as 2 efolder instances that run for the entire period of time between deploys. This has two downsides and maybe more. The first is that since our application steadily increases its memory utilization over its lifetime our application (specifically those two instances) will become less responsive over time. This is probably not a significant issue because traffic will be balanced across the rest of the newer instances as well so I don't imagine the responsiveness of our application will suffer all that significantly.

### Percent of total memory in use by EC2 instances
![image](https://user-images.githubusercontent.com/32683958/38106192-3ed80e70-335c-11e8-9025-ae6ccfdccaf8.png)
```
# Graph produced with the following query:
( 100 * ( node_memory_MemTotal{appname="efolder",env="prod",job="node-metrics",type="app"} - node_memory_MemFree{appname="efolder",env="prod",job="node-metrics",type="app"} ) ) / node_memory_MemTotal{appname="efolder",env="prod",job="node-metrics",type="app"}
```

The second downside is that if we have 2 long-running instances, they will not have gone through the deployment process recently and will not have received OS upgrades recently (since the deployment process is when we apply OS security updates). We could eliminate this risk by simply triggering a deployment every so often (once a week?). I am not sure how likely it is (well I kinda am, there is a 1 in 3.5 chance every night that an instance that survived the purge the previous night will survive the current night's purge... assuming AWS randomly decides which instances to kill off) or how much risk there would be in having an EC2 instance live for multiple days.

If my naive understanding of statistics is correct, the likelihood of there being EC2 instances that have lived longer than 2 days is very low (~4% chance there is an EC2 instance that has been alive for 3 days, 0.5% for 4 days, > 0.1% for 5 days), so I think there is very little risk here.

@evankroske Do you have any concerns about not deploying efolder instances frequently?

**Comment by lowellrex** [comment]
### Re: Ongoing time commitment for support staff
Before we consider the specifics of the corrupt zip file issue (or any other issue), I think it would be helpful to decide what that acceptable level of ongoing time commitment for support staff is so that we can weigh it against the price of a fix.

I agree that as more people use efolder express, we can expect to see more corrupt zip file issues (which will consume more support time), I just want to set a baseline on what that time commitment is.

**Comment by Chingujo** [comment]
We are agreed on the above so closing out this issue. Also added the note on eC2 instances at the bottom to not bury it in an issue. Thanks everyone!
